### ISSUE TITLE
Feature Request: binding for warp-ctc (Connectionist Temporal Classification loss)

### ISSUE THREAD

**Issue by jundengdeng** [pytorch/pytorch#570]
The Connectionist Temporal Classification (CTC) loss is very useful for automatic speech recognition. Baidu has implemented it and provided the binding for  tensorflow and torch (see [warp-ctc](https://github.com/baidu-research/warp-ctc)). It would be very helpful if pytorch would have also the binding for warp-ctc.

**Comment by SeanNaren** [comment]
I'll start having a look at this, seems like a similar interface to the Torch7 bindings would be suitable!

**Comment by SeanNaren** [comment]
Needs a review and a merge into the main warp-ctc library, but install instructions are [here](https://github.com/SeanNaren/warp-ctc/tree/pytorch_bindings/pytorch_binding). Example on how to use it is also given in the README!

I stress that if you have a pascal card you have to add a few flags into the CMakeLists.txt (god knows why it isn't merged already). Get flags [here](https://github.com/baidu-research/warp-ctc/pull/48).

Finally I've been working on a deepspeech package for PyTorch. If you are interested message me and I'll add you to the repo! Still WIP and will open once it plays nice :)

**Comment by jundengdeng** [comment]
thanks a lot for your effort. For sure, I'd like to contribute to a deepspeech model for PyTorch. Could you please add me to the repo? Thanks again!

**Comment by apaszke** [comment]
@SeanNaren do you want to merge it upstream as an additional pytorch binding? Wouldn't that be the best? You could just put all the code in `pytorch_binding` directory.

**Comment by SeanNaren** [comment]
@apaszke my bad should've mentioned I have a PR opened [here](https://github.com/baidu-research/warp-ctc/pull/71). I'd leave this ticket open till that's merged :)

**Comment by apaszke** [comment]
Ah, awesome! It looks really nice. I'd only consider changing the tests to use `unittest`.

**Comment by SeanNaren** [comment]
Thanks @apaszke! Just added unittest, just have the grads to do now :)

**Comment by apaszke** [comment]
It's already implemented so I'm closing the issue. Thanks @SeanNaren!

**Comment by meijieru** [comment]
Could not pass the unit test on my computer with CUDA enabled. The loss returns by gpu_ctc is always zero. @SeanNaren

**Comment by SeanNaren** [comment]
@meijieru to run the python tests for pytorch you have to run `python pytorch_binding/tests/test.py`. I'm not sure if the issue you having is specific to pytorch (as test_gpu is the interfaces' test)

**Comment by weedwind** [comment]
@SeanNaren I am trying to use the CTC loss, but I am not very clear how to fill the labels tensor. If I have three label sequences seq1, seq2, seq3, thay have different length L1, L2, L3, should the elements in the labels tensor follow the order [seq1[0:L1], seq2[0:L2], seq3[0:L3]], or [seq1[0], seq2[0], seq3[0], seq1[1], seq2[1], seq3[1]...]?  Also, I noticed that the labels never have zero? Is zero reserved for the blank label?

**Comment by LiangHao92** [comment]
@SeanNaren hello, I want to install warpctc binding pytorch. But I do not have a folder named "pytorch_binding", only a folder named "torch_binding". How can I binding pytorch now?
